tests/testsuite/package.rs

```diff
@@ -5038,3 +5038,303 @@ path = "examples/z.rs"
+  location searched: registry `crates-io`
+  required by package `level2 v0.0.1 ([CWD]/target/package/level2-0.0.1)`
+",
+        )
+        .run();
+}
+
+#[cargo_test]
+fn workspace_with_local_deps_packaging_one_with_needed_deps() {
+    let crates_io = registry::init();
+    let p = build_workspace_with_local_deps(None);
+
+    p.cargo("package -p level2 -p level3")
+        .replace_crates_io(crates_io.index_url())
+        .with_status(101)
+        .with_stdout("")
+        .with_stderr(
+            "\
+[PACKAGING] level2 v0.0.1 ([CWD]/level2)
+[PACKAGED] 3 files, [..] ([..] compressed)
+[VERIFYING] level2 v0.0.1 ([CWD]/level2)
+[UPDATING] crates.io index
+[ERROR] failed to verify package tarball
+
+Caused by:
+  no matching package named `level3` found
+  location searched: registry `crates-io`
+  required by package `level2 v0.0.1 [..]
+",
+        )
+        .run();
+}
+
+#[cargo_test]
+fn workspace_with_local_deps_list() {
+    let crates_io = registry::init();
+    let p = build_workspace_with_local_deps(None);
+
+    p.cargo("package --list")
+        .replace_crates_io(crates_io.index_url())
+        .with_stdout(
+            // This looks like it could be improved. See
+            // https://github.com/rust-lang/cargo/issues/13953
+            "\
+Cargo.lock
+Cargo.toml
+Cargo.toml.orig
+src/main.rs
+Cargo.toml
+Cargo.toml.orig
+src/lib.rs
+Cargo.toml
+Cargo.toml.orig
+src/lib.rs
+",
+        )
+        .with_stderr("")
+        .run();
+}
+
+#[cargo_test]
+fn workspace_with_local_deps_index_mismatch() {
+    let alt_reg = registry::RegistryBuilder::new()
+        .http_api()
+        .http_index()
+        .alternative()
+        .build();
+    // We're publishing to an alternate index, but the manifests don't specify it.
+    // The intra-workspace deps won't be found.
+    let p = build_workspace_with_local_deps(None);
+    p.cargo(&format!("package --index {}", alt_reg.index_url()))
+        .with_status(1)
+        .with_stdout("")
+        .with_stderr(
+            "\
+error: unexpected argument '--index' found
+
+Usage: cargo package [OPTIONS]
+
+For more information, try '--help'.
+",
+        )
+        .run();
+}
+
+#[cargo_test]
+fn workspace_with_local_deps_alternative_index() {
+    let alt_reg = registry::RegistryBuilder::new()
+        .http_api()
+        .http_index()
+        .alternative()
+        .build();
+    let p = build_workspace_with_local_deps(Some("alternative"));
+
+    p.cargo(&format!("package --index {}", alt_reg.index_url()))
+        .with_status(1)
+        .with_stdout("")
+        .with_stderr(
+            "\
+error: unexpected argument '--index' found
+
+Usage: cargo package [OPTIONS]
+
+For more information, try '--help'.
+",
+        )
+        .run();
+}
+
+/// level1 -> level2 checks that we handle local dependencies.
+/// level2 -> level3 checks that we handle dependencies with local dependencies.
+fn build_workspace_with_local_deps(registry: Option<&str>) -> Project {
```

Comment:
Generally not too big of a fan of fixture setups like this
- The abstraction is leaky: in reading the test, you have to know all of the details of this to make sense of but but this has bad code locality
- This tends towards "god tests" that cover too many cases.  There are times when a test needs to hit a bunch of cases at once but when you have helpers like this, they tend to get over-reused and every test hits every case when only a few need to.  This causes slower tests that are likely to hit storage limits (a regular problem for cargo's tests on CI)